File: foot-match-app/services/authService.ts
import { request, setTokens, clearTokens, getRefreshToken } from './api';
import type { AuthResponse, LoginRequest, RegisterRequest } from '@/types';

export const authService = {
  async login(data: LoginRequest): Promise<AuthResponse> {
    const response = await request<AuthResponse>('/api/auth/login', {
      method: 'POST',
      body: JSON.stringify(data),
    });
    await setTokens(response.accessToken, response.refreshToken);
    return response;
  },

  async register(data: RegisterRequest): Promise<AuthResponse> {
    const response = await request<AuthResponse>('/api/auth/register', {
      method: 'POST',
      body: JSON.stringify(data),
    });
    await setTokens(response.accessToken, response.refreshToken);
    return response;
  },

  async refresh(): Promise<AuthResponse | null> {
    const refreshToken = await getRefreshToken();
    if (!refreshToken) return null;

    try {
      const response = await request<AuthResponse>('/api/auth/refresh', {
        method: 'POST',
        body: JSON.stringify({ refreshToken }),
      });
      await setTokens(response.accessToken, response.refreshToken);
      return response;
    } catch {
      await clearTokens();
      return null;
    }
  },

  async logout(): Promise<void> {
    await clearTokens();
  },
};

File: foot-match-api/src/main/java/pl/pzynis/footmatch/application/service/AuthService.java
package pl.pzynis.footmatch.application.service;

import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import pl.pzynis.footmatch.api.dto.AuthResponse;
import pl.pzynis.footmatch.api.dto.LoginRequest;
import pl.pzynis.footmatch.api.dto.RegisterRequest;
import pl.pzynis.footmatch.api.dto.UserResponse;
import pl.pzynis.footmatch.domain.exception.EmailAlreadyExistsException;
import pl.pzynis.footmatch.domain.exception.InvalidCredentialsException;
import pl.pzynis.footmatch.infrastructure.persistence.entity.UserEntity;
import pl.pzynis.footmatch.infrastructure.persistence.repository.UserRepository;
import pl.pzynis.footmatch.infrastructure.security.JwtTokenProvider;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider jwtTokenProvider;

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        if (userRepository.existsByEmail(request.email())) {
            throw new EmailAlreadyExistsException(request.email());
        }

        UserEntity user = UserEntity.builder()
                .email(request.email())
                .passwordHash(passwordEncoder.encode(request.password()))
                .name(request.name())
                .role(request.role())
                .build();

        user = userRepository.save(user);

        return generateAuthResponse(user);
    }

    public AuthResponse login(LoginRequest request) {
        UserEntity user = userRepository.findByEmail(request.email())
                .orElseThrow(() -> new InvalidCredentialsException());

        if (!passwordEncoder.matches(request.password(), user.getPasswordHash())) {
            throw new InvalidCredentialsException();
        }

        return generateAuthResponse(user);
    }

    public AuthResponse refreshToken(String refreshToken) {
        if (!jwtTokenProvider.validateToken(refreshToken)) {
            throw new InvalidCredentialsException();
        }

        var userId = jwtTokenProvider.getUserIdFromToken(refreshToken);
        UserEntity user = userRepository.findById(userId)
                .orElseThrow(() -> new InvalidCredentialsException());

        return generateAuthResponse(user);
    }

    private AuthResponse generateAuthResponse(UserEntity user) {
        String accessToken = jwtTokenProvider.generateAccessToken(
                user.getId(),
                user.getEmail(),
                user.getRole().name()
        );
        String refreshToken = jwtTokenProvider.generateRefreshToken(
                user.getId(),
                user.getEmail(),
                user.getRole().name()
        );

        UserResponse userResponse = new UserResponse(
                user.getId(),
                user.getEmail(),
                user.getName(),
                user.getRole()
        );

        return new AuthResponse(
                accessToken,
                refreshToken,
                jwtTokenProvider.getAccessTokenExpiration() / 1000,
                userResponse
        );
    }
}
